None value in `array` decoder raise `TypeError` exception

* PyDecoder version: 1.0.0
* Python version: ANY
* Operating System: ANY

### Description

When `vals` param in `array` function is `None`, then exception is raised, instead returning error result.

### What I Did

```python
from pydecoder.structs import array
from pydecoder.json import to_string

array(to_string, None)


    @curry
    def array(factory, vals):
        '''Decode string/value as list'''
        result = []

>       for val in vals:
E       TypeError: 'NoneType' object is not iterable
```

--- a/pydecoder/structs.py
+++ b/pydecoder/structs.py
@@ -1,13 +1,26 @@
 # -*- coding: utf-8 -*-
 '''Decoder structures'''
 
-from pyresult import is_error, value, ok
+from pyresult import is_error, value, ok, error
 from toolz import curry
 
 
 @curry
 def array(factory, vals):
-    '''Decode string/value as list'''
+    '''Decode string/value as list
+
+    array :: (a -> Result e b) -> List a -> Result e (List b)
+
+    >>> from pydecoder.primitives import to_int
+    >>> array(to_int, [1, 2, 3])
+    Result(status='Ok', value=[1, 2, 3])
+
+    >>> array(to_int, None)
+    Result(status='Error', value="'None' isn't list or tuple.")
+    '''
+    if not isinstance(vals, (list, tuple)):
+        return error('\'{}\' isn\'t list or tuple.'.format(vals))
+
     result = []
 
     for val in vals:

--- a/pydecoder/structs.py
+++ b/pydecoder/structs.py
@@ -2,6 +2,7 @@
 '''Decoder structures'''
 
 from pyresult import is_error, value, ok, error
+from six import u
 from toolz import curry
 
 
@@ -19,7 +20,7 @@ def array(factory, vals):
     Result(status='Error', value="'None' isn't list or tuple.")
     '''
     if not isinstance(vals, (list, tuple)):
-        return error('\'{}\' isn\'t list or tuple.'.format(vals))
+        return error(u('\'{}\' isn\'t list or tuple.').format(vals))
 
     result = []
 

--- a/setup.cfg
+++ b/setup.cfg
@@ -23,14 +23,14 @@ exclude = urls.py,runtests.py,conf.py
 ignore = E501,E402
 
 [tool:pytest]
-addopts = -vvl
+addopts = -vvl --doctest-modules --ignore=setup.py
 norecursedirs = .* requirements docs node_modules
 
 [coverage:run]
 branch = true
 
 [coverage:report]
-omit = 
+omit =
 	*site-packages*
 	*tests*
 	*.tox*

--- a/tests/test_structs.py
+++ b/tests/test_structs.py
@@ -5,12 +5,37 @@ from pyresult import is_ok, is_error
 from six import string_types
 from six.moves import zip
 
+from hypothesis import given
+from hypothesis.strategies import (
+    lists,
+    one_of,
+    none,
+    integers,
+    recursive,
+    text,
+    booleans,
+    floats,
+    complex_numbers,
+    dictionaries
+)
+
 from pydecoder.primitives import to_int
 from pydecoder.structs import array
 
 
-def test_array_returns_decoded_list_in_result():
-    ins = ['1', '2', '3']
+NOT_LIST = recursive(
+    none() |
+    text() |
+    booleans() |
+    integers() |
+    floats() |
+    complex_numbers(),
+    lambda children: dictionaries(text(), children)
+)
+
+
+@given(lists(integers()))
+def test_array_returns_decoded_list_in_result(ins):
 
     rv = array(to_int, ins)
 
@@ -20,10 +45,17 @@ def test_array_returns_decoded_list_in_result():
         assert res == int(orig)
 
 
-def test_stop_on_first_error_and_returns_it():
-    ins = ['1', None, '3']
-
+@given(lists(one_of(integers(), none()), min_size=1).filter(lambda x: None in x))
+def test_stop_on_first_error_and_returns_it(ins):
     rv = array(to_int, ins)
+
+    assert is_error(rv)
+    assert isinstance(rv.value, string_types)
+
+
+@given(NOT_LIST)
+def test_array_returns_error_if_value_isnt_valid_list(val):
+    rv = array(to_int, val)
 
     assert is_error(rv)
     assert isinstance(rv.value, string_types)

